File: apps/dot-net-sdk/Firecrawl/Models/MenuProfile.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Firecrawl.Models;

/// <summary>
/// Structured menu information extracted via the <c>menu</c> scrape format.
/// </summary>
public class MenuProfile
{
    [JsonPropertyName("isMenu")]
    public bool IsMenu { get; set; }

    [JsonPropertyName("confidence")]
    public double Confidence { get; set; }

    [JsonPropertyName("merchant")]
    public MenuMerchant Merchant { get; set; } = new();

    [JsonPropertyName("currency")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Currency { get; set; }

    [JsonPropertyName("sections")]
    public List<MenuSection> Sections { get; set; } = new();

    [JsonPropertyName("sourceUrl")]
    public string SourceUrl { get; set; } = string.Empty;
}

/// <summary>
/// The merchant a menu belongs to.
/// </summary>
public class MenuMerchant
{
    [JsonPropertyName("name")]
    public string Name { get; set; } = string.Empty;

    [JsonPropertyName("type")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Type { get; set; }

    [JsonPropertyName("location")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public JsonElement? Location { get; set; }
}

/// <summary>
/// An ordered section of a menu (e.g. "Appetizers"), holding items.
/// </summary>
public class MenuSection
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = string.Empty;

    [JsonPropertyName("name")]
    public string Name { get; set; } = string.Empty;

    [JsonPropertyName("description")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Description { get; set; }

    [JsonPropertyName("items")]
    public List<MenuItem> Items { get; set; } = new();
}

/// <summary>
/// A single item on a menu, with its own pricing, availability, and images.
/// </summary>
public class MenuItem
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = string.Empty;

    [JsonPropertyName("name")]
    public string Name { get; set; } = string.Empty;

    [JsonPropertyName("description")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Description { get; set; }

    [JsonPropertyName("images")]
    public List<MenuImage> Images { get; set; } = new();

    [JsonPropertyName("price")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public MenuPrice? Price { get; set; }

    [JsonPropertyName("availability")]
    public MenuAvailability Availability { get; set; } = new();

    [JsonPropertyName("dietary")]
    public List<string> Dietary { get; set; } = new();

    [JsonPropertyName("calories")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public double? Calories { get; set; }

    [JsonPropertyName("optionGroups")]
    public List<object> OptionGroups { get; set; } = new();

    [JsonPropertyName("identifiers")]
    public MenuItemIdentifiers Identifiers { get; set; } = new();

    [JsonPropertyName("url")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Url { get; set; }

    [JsonPropertyName("sourceUrl")]
    public string SourceUrl { get; set; } = string.Empty;
}

/// <summary>
/// An image associated with a menu item.
/// </summary>
public class MenuImage
{
    [JsonPropertyName("url")]
    public string Url { get; set; } = string.Empty;

    [JsonPropertyName("alt")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Alt { get; set; }
}

/// <summary>
/// A monetary price for a menu item.
/// </summary>
public class MenuPrice
{
    [JsonPropertyName("amount")]
    public double Amount { get; set; }

    [JsonPropertyName("currency")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Currency { get; set; }

    [JsonPropertyName("formatted")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Formatted { get; set; }
}

/// <summary>
/// Availability information for a menu item.
/// </summary>
public class MenuAvailability
{
    [JsonPropertyName("inStock")]
    public bool InStock { get; set; }

    [JsonPropertyName("text")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Text { get; set; }
}

/// <summary>
/// External identifiers for a menu item.
/// </summary>
public class MenuItemIdentifiers
{
    [JsonPropertyName("merchantItemId")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? MerchantItemId { get; set; }
}

File: apps/ruby-sdk/lib/firecrawl/models/menu_profile.rb
# frozen_string_literal: true

module Firecrawl
  module Models
    # Structured menu information extracted from a restaurant/merchant page
    # via the `menu` scrape format.
    class MenuProfile
      # An image associated with a menu item.
      class Image
        attr_reader :url, :alt

        def initialize(data)
          @url = data["url"]
          @alt = data["alt"]
        end
      end

      # A monetary value with an optional currency and formatted string.
      class Price
        attr_reader :amount, :currency, :formatted

        def initialize(data)
          @amount = data["amount"]
          @currency = data["currency"]
          @formatted = data["formatted"]
        end
      end

      # Stock availability information for a menu item. Always present.
      class Availability
        attr_reader :in_stock, :text

        def initialize(data)
          data ||= {}
          @in_stock = data["inStock"] || false
          @text = data["text"]
        end
      end

      # Merchant (restaurant/business) profile for the menu.
      class Merchant
        attr_reader :name, :type, :location

        def initialize(data)
          data ||= {}
          @name = data["name"]
          @type = data["type"]
          @location = data["location"]
        end
      end

      # Identifiers carried on a menu item.
      class Identifiers
        attr_reader :merchant_item_id

        def initialize(data)
          data ||= {}
          @merchant_item_id = data["merchantItemId"]
        end
      end

      # A single item on the menu. Pricing, availability, images, and dietary
      # information live here rather than on the section or profile.
      class Item
        attr_reader :id, :name, :description, :images, :price, :availability,
                    :dietary, :calories, :option_groups, :identifiers, :url,
                    :source_url

        def initialize(data)
          @id = data["id"]
          @name = data["name"]
          @description = data["description"]
          @images = (data["images"] || []).map { |img| Image.new(img) }
          @price = data["price"] && Price.new(data["price"])
          @availability = Availability.new(data["availability"])
          @dietary = data["dietary"] || []
          @calories = data["calories"]
          @option_groups = data["optionGroups"] || []
          @identifiers = Identifiers.new(data["identifiers"])
          @url = data["url"]
          @source_url = data["sourceUrl"]
        end
      end

      # An ordered group of menu items.
      class Section
        attr_reader :id, :name, :description, :items

        def initialize(data)
          @id = data["id"]
          @name = data["name"]
          @description = data["description"]
          @items = (data["items"] || []).map { |item| Item.new(item) }
        end
      end

      attr_reader :is_menu, :confidence, :merchant, :currency, :sections,
                  :source_url

      def initialize(data)
        @is_menu = data["isMenu"] || false
        @confidence = data["confidence"]
        @merchant = Merchant.new(data["merchant"])
        @currency = data["currency"]
        @sections = (data["sections"] || []).map { |section| Section.new(section) }
        @source_url = data["sourceUrl"]
      end

      def to_s
        "MenuProfile{merchant=#{merchant&.name || 'unknown'}, sourceUrl=#{source_url || 'unknown'}}"
      end
    end
  end
end
